mantidproject/mantid
# Set transmission runs based on run angle

This PR adds a table to the Settings tab of the ISIS Reflectometry GUI which allows default transmission runs to be specified on a per-angle basis, so that the transmission runs can be looked up based on the angle of each run.

This replaces the existing text entry box which just allowed for one default set of transmission runs to be specified. Overall defaults can still be specified by entering one row where the angle column is blank. Due to the issues with providing the lookup angle described in the linked issue, for now, to avoid confusion, there is a restriction that per-angle OR overall defaults can be specified, but not both.

This PR also includes the following changes as part of an effort to make the underlying code more maintainable:
- Significant refactoring to cache row metadata, such as properties used in the reduction, the processed state, the reduced workspace names etc. This information was previously being recalculated in many different places whenever required but it was becoming increasingly difficult and error-prone to keep track of all of the information required to recalculate these values. These changes should make it easier going forward to e.g. change the naming convention for workspace slices.
- The `Output Notebook` option now works for all groups that are processed as non-event workspaces. Previously, if event handling was enabled but a group contained non-event workspaces, generating the notebook was being silently skipped.
- Plotting rows in event handling mode now plots the `IvsQ_binned_` workspaces rather than `IvsQ_`, to make the behaviour consistent with non-event mode.

Fixes #20088

**To test:**
Check that the `ISIS Reflectometry` and `ISIS SANS v2 experimental` GUIs still work as expected.

On the reflectometry GUI:
- Try specifying defaults with and/or without angles.
- Try overriding transmission runs in the table on the Runs tab - this should take precedence over anything on the Settings tab.
- Check that other instrument/experiment properties are still picked up ok from the Settings tab and that these also get overridden by the properties on the Runs tab.
- Try overriding properties in the Runs tab using both the normal columns (e.g. `dQ/Q`) as well as the `Options` column (e.g. `WavelengthMax` in here should override `LambdaMax` on the Settings tab).
- Check that properties and defaults are passed through ok to the preprocessing and postprocessing algorithms (i.e. `CreateTransmissionWorkspaceAuto` and `Stitch1DMany`).
- Check that the plotting buttons on the GUI work ok.
- Check that `Output Notebook` still works and that the code generated still runs ok. Check that the plots are the same as those from the GUI (although note that they are shown with error bars in the notebook and without from the GUI. Note also that there's an existing bug where the Params must be set in the stitch properties before first running the reduction because changes aren't picked up.)

---

#### Reviewer ####

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):

##### Code Review #####


##### Functional Tests #####


Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.
